File: java/2326. Spiral Matrix IV.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[][] spiralMatrix(int m, int n, ListNode head) {
        int[][] result = new int[m][n];
        for (int[] i : result) Arrays.fill(i, -1);
        int top = -1, left = -1, right = n, bottom = m;
        int dir = 0;
        int x = 0, y = 0;
        while (head != null) {
            result[x][y] = head.val;
            head = head.next;
            if (dir == 0) {
                y++;
                if (y == right) {
                    y--;
                    x++;
                    dir = 1;
                    top++;
                }
            } else if (dir == 1) {
                x++;
                if (x == bottom) {
                    x--;
                    y--;
                    dir = 2;
                    right--;
                }
            } else if (dir == 2) {
                y--;
                if (y == left) {
                    y++;
                    x--;
                    dir = 3;
                    bottom--;
                }
            } else {
                x--;
                if (x == top) {
                    x++;
                    y++;
                    dir = 0;
                    left++;
                }
            }
        }
        return result;
    }
}

File: go/2326. Spiral Matrix IV.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func spiralMatrix(m int, n int, head *ListNode) [][]int {
    result := make([][]int, m)
    for i := range result {
        result[i] = make([]int, n)
        for j := range result[i] {
            result[i][j] = -1
        }
    }
    top, left, right, bottom, dir := -1, -1, n, m, 0
    x, y := 0, 0
    for head != nil {
        result[x][y] = head.Val
        head = head.Next
        if dir == 0 {
            y++
            if y == right {
                x, y, dir, top = x+1, y-1, 1, top+1
            }
        } else if dir == 1 {
            x++
            if x == bottom {
                x, y, dir, right = x-1, y-1, 2, right-1
            }
        } else if dir == 2 {
            y--
            if y == left {
                x, y, dir, bottom = x-1, y+1, 3, bottom-1
            }
        } else {
            x--
            if x == top {
                x, y, dir, left = x+1, y+1, 0, left+1
            }
        }
    }
    return result
}
